### ISSUE TITLE
The mAP on val2017 for the posted model

### ISSUE THREAD

**Issue by yizhouyan** [yhenon/pytorch-retinanet#63]
I tried the posted model on the COCO val2017. However, I got the following results:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.099
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.165
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.099
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.047
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.102
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.151
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.111
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.161
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.170
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.078
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.177
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.242

The results are generated by:
dataset_val = CocoDataset("../COCO/", set_name='val2017', transform=transforms.Compose([Normalizer(), Resizer()]))
retinanet = torch.load("coco_resnet_50_map_0_335.pt")
coco_eval.evaluate_coco(dataset_val, retinanet)

Anyone has tried the posted model on COCO val2017? Any ideas on why the mAP is so low?

Thanks!

**Comment by IgorSusmelj** [comment]
I had similar issues. I'm currently retraining the model myself with the provided weights as initialization.

After one epoch I ended up having those results. Especially for small objects this model seems to have some issues...

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.241
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.333
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.269
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.060
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.291
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.391
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.202
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.262
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.263
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.060
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.309
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.433

**Comment by IgorSusmelj** [comment]
**Update**: After training a resnet-50 for several day on a 1080ti I get the following scores on Coco val 2017

Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.300
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.470
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.317
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.133
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.329
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.420
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.260
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.396
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.425
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.219
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.474
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.552

**Comment by distillation-dcf** [comment]
@IgorSusmelj Did you modify the original train.py in the repo during your training process? And how many epochs have you trained for achieving mAP of 0.30?

**Comment by IgorSusmelj** [comment]
I only adapted the code to run on a single GPU instead of 8. To get mAP of 0.3 I trained for 3-4 days. I think that was around 10 epochs.

**Comment by distillation-dcf** [comment]
> I only adapted the code to run on a single GPU instead of 8. To get mAP of 0.3 I trained for 3-4 days. I think that was around 10 epochs.

@IgorSusmelj Similarly, I remove ```torch.nn.DataParallel``` in train.py and train on a single TITAN X with all the other default settings. But I only get mAP of 0.16 after 1 epoch and 0.26 after 10 epochs. Did you add or modify any other things, like data augmentation or use other optimization with other learning rate?

**Comment by westnight** [comment]
I ran this code on another dataset with 2 1080ti gpus, and performance is not good as you gays. I think something's wrong with the code.

**Comment by huangzicheng** [comment]
![2019-08-28 14-59-33屏幕截图](https://user-images.githubusercontent.com/30286349/63833314-4ec79400-c9a5-11e9-9092-e054df411770.png)
it same the code  not right

**Comment by etienne87** [comment]
are we sure the resizer is not changing slightly the coordinates? i'm using the dataloader with batchsize > 1 and images are padded, so boxes are different. However coco_eval uses the original val2017.json as ground truth...

**Comment by xiaofenghuhuhu** [comment]
> are we sure the resizer is not changing slightly the coordinates? i'm using the dataloader with batchsize > 1 and images are padded, so boxes are different. However coco_eval uses the original val2017.json as ground truth...

Maybe it's because the reproduction of this code does not use bidirectional linear interpolation to process the image size. You can look at (e) in Table 1 in the paper and things will become clear.

And here is my No.10 epoch's result, for batch_size=4, and DataParallel available

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.293
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.461
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.310
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.125
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.325
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.415
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.256
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.391
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.418
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.208
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.465
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.554

**Comment by yhenon** [comment]
I've just repeated the validation and get the following results:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.335
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.499
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.357
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.167
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.369
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.466
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.282
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.429
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.458
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.255
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.508
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.597
```

Ensure you are resizing properly (keep aspect ratio, min image size of 608 but no greater than 1024), and performing per-channel mean/standardization properly.

**Comment by yhenon** [comment]
I have added a `coco_validation.py` which reproduces the above results.

**Comment by zjysteven** [comment]
> Ensure you are resizing properly (keep aspect ratio, min image size of 608 but no greater than 1024)

@yhenon May I ask why you choose these two specific min/max sizes? I'm asking as in the original paper they used 600*600 as their input size, and in keras-retinanet, they use 800 and 1333 as the min and max size respectively. 

I know that these arguments should be dependent on the dataset, but I'm just confused about how you choosed these default values (608 and 1024)? Thanks!
